{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# Tutorial: 10x multiome pbmc\n",
    "\n",
    "\n",
    "The data consists of *PBMC from a Healthy Donor - Granulocytes Removed Through Cell Sorting (3k)* which is freely available from 10x Genomics (click [here](https://www.10xgenomics.com/resources/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-3-k-1-standard-2-0-0), some personal information needs to be provided before you can gain access to the data). This is a multi-ome dataset.\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**Note:**\n",
    "\n",
    "In this notebook we will only show the minimal steps needed for running the SCENIC+ analysis. For more information on analysing scRNA-seq data and scATAC-seq data we refer the reader to other tutorials (e.g. [Scanpy](https://scanpy-tutorials.readthedocs.io/en/latest/index.html) and [pycisTopic](https://pycistopic.readthedocs.io/en/latest/) in python or [Seurat](https://satijalab.org/seurat/) and [cisTopic](https://github.com/aertslab/cisTopic) or [Signac](https://satijalab.org/signac/) in R).\n",
    "\n",
    "</div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set-up environment and download data \n",
    "We will first create a directory to store the data and results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pycisTopic'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mos\u001b[39;00m\n\u001b[1;32m      2\u001b[0m work_dir \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mpbmc_tutorial\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m----> 3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpycisTopic\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m#set some figure parameters for nice display inside jupyternotebooks.\u001b[39;00m\n\u001b[1;32m      5\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39mrun_line_magic(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmatplotlib\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124minline\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'pycisTopic'"
     ]
    }
   ],
   "source": [
    "import os\n",
    "work_dir = 'pbmc_tutorial'\n",
    "import pycisTopic\n",
    "#set some figure parameters for nice display inside jupyternotebooks.\n",
    "%matplotlib inline\n",
    "\n",
    "#make a directory for to store the processed scRNA-seq data.\n",
    "if not os.path.exists(os.path.join(work_dir, 'scATAC')):\n",
    "    os.makedirs(os.path.join(work_dir, 'scATAC'))\n",
    "tmp_dir = '/oscar/data/rsingh47/tpham43/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Specify the location of the ATAC fragments file, this is the main input into pycisTopic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "fragments_dict = {'10x_pbmc': os.path.join(work_dir, 'data/pbmc_granulocyte_sorted_3k_atac_fragments.tsv.gz')}   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate pseudobulk ATAC-seq profiles, call peaks and generate a consensus peak set\n",
    "\n",
    "We will use the cell type labels from the scRNA-seq side of the data to generate pseudobulk profiles per cell type. These pseudobulk profiles will be used to call peaks which will be merged in one consensus peak set.\n",
    "\n",
    "The advantage of calling peaks per cell type (instead of calling peaks on the whole bulk profile at once) is that, for rare cell types, there is a risk that the ATAC signal of these cells might be confused for noise by the peak caller when viewed in the whole bulk profile. Calling peaks per cell type helps resolving these rare peaks and thus increases the resolution of the data.\n",
    "\n",
    "We will first load the cell type annotation we generated in the scRNA-seq analysis above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2603 × 340\n",
       "    obs: 'n_genes', 'doublet_score', 'predicted_doublet', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'ingest_celltype_label', 'leiden_res_0.8', 'celltype'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'mean', 'std'\n",
       "    uns: 'celltype_colors', 'hvg', 'ingest_celltype_label_colors', 'leiden', 'leiden_res_0.8_colors', 'log1p', 'neighbors', 'pca', 'scrublet', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap', 'rep'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "adata = sc.read_h5ad(os.path.join(work_dir, 'scRNA/adata.h5ad'))\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-0.0253034 ,  0.00465935, -0.01765437, ...,  0.03878163,\n",
       "         0.05171565, -0.09465852],\n",
       "       [-0.01848559,  0.00251321,  0.00181723, ...,  0.00191663,\n",
       "         0.01899515,  0.02849598],\n",
       "       [-0.00334583,  0.02559683,  0.01415306, ...,  0.03252344,\n",
       "        -0.07208318, -0.03594767],\n",
       "       ...,\n",
       "       [ 0.1340777 ,  0.02612962, -0.02800387, ..., -0.03469239,\n",
       "        -0.0420995 , -0.01510095],\n",
       "       [ 0.00076975, -0.07723995,  0.10534835, ..., -0.11510067,\n",
       "         0.05080847, -0.00559565],\n",
       "       [-0.02409383, -0.00278305, -0.01773829, ...,  0.07455043,\n",
       "        -0.07981651, -0.0718854 ]])"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.varm['PCs']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "pca = pd.DataFrame(abs(adata.varm['PCs']),index =adata.var.index ).T\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "variance = adata.uns['pca']['variance']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "weighted_feature = pd.DataFrame(pca.T @ variance, columns = ['Weight']).sort_values(by=\"Weight\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['FCER1G', 'HLA-DQA1', 'CCL5', 'HLA-DRB1', 'GZMB', 'ARHGAP24', 'GZMH',\n",
       "       'TYROBP', 'HLA-DPB1', 'GZMA'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_10_genes = weighted_feature.index[:10]\n",
    "top_10_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "cistopic_obj = pickle.load(open(os.path.join(work_dir, 'scATAC/cistopic_obj.pkl'), 'rb'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_matrix = cistopic_obj.fragment_matrix.toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_peaks = pd.DataFrame(cistopic_obj.region_names)\n",
    "atac_peaks.rename(columns={0: 'features'}, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_barcode= pd.DataFrame(cistopic_obj.cell_names)\n",
    "atac_barcode.rename(columns={0: 'orig.ident'}, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/oscar/data/koconno5/tpham43/env_scenicplus/lib64/python3.9/site-packages/anndata/_core/aligned_df.py:67: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2412 × 194115\n",
       "    obs: 'orig.ident'\n",
       "    var: 'features'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### atac anndata\n",
    "atac_adata = sc.AnnData(atac_matrix.T,\n",
    "                  var = atac_peaks['features'].to_frame(),\n",
    "                  obs = atac_barcode['orig.ident'].to_frame(),\n",
    "#                       var_names='gene_symbols'                # use gene symbols for the variable names (variables-axis index\n",
    "                  )\n",
    "atac_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>101175</td>\n",
       "      <td>101675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>101908</td>\n",
       "      <td>102408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GL000194.1</td>\n",
       "      <td>114713</td>\n",
       "      <td>115213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>22453</td>\n",
       "      <td>22953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GL000195.1</td>\n",
       "      <td>32234</td>\n",
       "      <td>32734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194110</th>\n",
       "      <td>chrX</td>\n",
       "      <td>40170588</td>\n",
       "      <td>40171088</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194111</th>\n",
       "      <td>chrX</td>\n",
       "      <td>45768071</td>\n",
       "      <td>45768571</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194112</th>\n",
       "      <td>chrX</td>\n",
       "      <td>3711862</td>\n",
       "      <td>3712362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194113</th>\n",
       "      <td>chrX</td>\n",
       "      <td>107715002</td>\n",
       "      <td>107715502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194114</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154370450</td>\n",
       "      <td>154370950</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>194115 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chromosome      Start        End\n",
       "0       GL000194.1     101175     101675\n",
       "1       GL000194.1     101908     102408\n",
       "2       GL000194.1     114713     115213\n",
       "3       GL000195.1      22453      22953\n",
       "4       GL000195.1      32234      32734\n",
       "...            ...        ...        ...\n",
       "194110        chrX   40170588   40171088\n",
       "194111        chrX   45768071   45768571\n",
       "194112        chrX    3711862    3712362\n",
       "194113        chrX  107715002  107715502\n",
       "194114        chrX  154370450  154370950\n",
       "\n",
       "[194115 rows x 3 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_peaks[['Chromosome', 'Start','End']] = atac_peaks['features'].str.split(':|-', expand=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_peaks[['Chromosome', 'Start','End']].to_csv(\"./atac_peaks.bed\",header = False, index = False, sep = \"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/bin/bash: line 1: bedtools: command not found\r\n"
     ]
    }
   ],
   "source": [
    "# !module load bedtools2\n",
    "# !bedtools window -w 2000 -a ./gencode.v38.annotation.gff3.gz -b atac_peaks.bed > atac_peaks_filtered.bed\n",
    "# !cat atac_peaks_filtered.bed | cut -f4,5,6,7 > atac_peaks_filtered_sub.bed\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def overlap(q_st, q_end, res_st, res_end):\n",
    "    o  = min(q_end, res_end)-max(q_st, res_st)\n",
    "    return o\n",
    "def gencode_all_known_genes(a, tb):\n",
    "    genes = []\n",
    "    try:\n",
    "        for region in tb.fetch(a['chr'], int(a['start']), int(a['end'])):\n",
    "            if region:\n",
    "                r = region.split('\\t')\n",
    "                overlap_len = overlap(int(a['start']), int(a['end']), int(r[1]), int(r[2]))\n",
    "                ret_val = '{}({})'.format(r[3], np.round(overlap_len/float(int(a['end'])-int(a['start']))*100, 2)) ### Percentage of the input interval that overlap with the gene\n",
    "                genes.append(ret_val)\n",
    "        if len(genes)>0:\n",
    "            return \";\".join(genes)\n",
    "        else:\n",
    "            return \"NA(0)\"\n",
    "    except ValueError:\n",
    "        return \"NA(0)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>NA1</th>\n",
       "      <th>NA2</th>\n",
       "      <th>NA3</th>\n",
       "      <th>attribute</th>\n",
       "      <th>seqname_2</th>\n",
       "      <th>start_2</th>\n",
       "      <th>end_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>chr1</td>\n",
       "      <td>817073</td>\n",
       "      <td>817573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>chr1</td>\n",
       "      <td>822183</td>\n",
       "      <td>822683</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>chr1</td>\n",
       "      <td>819940</td>\n",
       "      <td>820440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>chr1</td>\n",
       "      <td>821647</td>\n",
       "      <td>822147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>586071</td>\n",
       "      <td>827796</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>chr1</td>\n",
       "      <td>825291</td>\n",
       "      <td>825791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4179780</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>155889458</td>\n",
       "      <td>155889547</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000460621.6:2;Parent=ENST0000046...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155891359</td>\n",
       "      <td>155891859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4179781</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>155889467</td>\n",
       "      <td>155889547</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=CDS:ENST00000460621.6;Parent=ENST0000046062...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155891359</td>\n",
       "      <td>155891859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4179782</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>155889467</td>\n",
       "      <td>155889469</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=start_codon:ENST00000460621.6;Parent=ENST00...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155891359</td>\n",
       "      <td>155891859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4179783</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>155881378</td>\n",
       "      <td>155881448</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR5:ENST00000460621.6;Parent=ENST000004606...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155881049</td>\n",
       "      <td>155881549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4179784</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>155889458</td>\n",
       "      <td>155889466</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=UTR5:ENST00000460621.6;Parent=ENST000004606...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>155891359</td>\n",
       "      <td>155891859</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4179785 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname  source         feature      start        end NA1 NA2 NA3  \\\n",
       "0          chr1  HAVANA            gene     586071     827796   .   -   .   \n",
       "1          chr1  HAVANA            gene     586071     827796   .   -   .   \n",
       "2          chr1  HAVANA            gene     586071     827796   .   -   .   \n",
       "3          chr1  HAVANA            gene     586071     827796   .   -   .   \n",
       "4          chr1  HAVANA            gene     586071     827796   .   -   .   \n",
       "...         ...     ...             ...        ...        ...  ..  ..  ..   \n",
       "4179780    chrX  HAVANA            exon  155889458  155889547   .   +   .   \n",
       "4179781    chrX  HAVANA             CDS  155889467  155889547   .   +   0   \n",
       "4179782    chrX  HAVANA     start_codon  155889467  155889469   .   +   0   \n",
       "4179783    chrX  HAVANA  five_prime_UTR  155881378  155881448   .   +   .   \n",
       "4179784    chrX  HAVANA  five_prime_UTR  155889458  155889466   .   +   .   \n",
       "\n",
       "                                                 attribute seqname_2  \\\n",
       "0        ID=ENSG00000230021.10;gene_id=ENSG00000230021....      chr1   \n",
       "1        ID=ENSG00000230021.10;gene_id=ENSG00000230021....      chr1   \n",
       "2        ID=ENSG00000230021.10;gene_id=ENSG00000230021....      chr1   \n",
       "3        ID=ENSG00000230021.10;gene_id=ENSG00000230021....      chr1   \n",
       "4        ID=ENSG00000230021.10;gene_id=ENSG00000230021....      chr1   \n",
       "...                                                    ...       ...   \n",
       "4179780  ID=exon:ENST00000460621.6:2;Parent=ENST0000046...      chrX   \n",
       "4179781  ID=CDS:ENST00000460621.6;Parent=ENST0000046062...      chrX   \n",
       "4179782  ID=start_codon:ENST00000460621.6;Parent=ENST00...      chrX   \n",
       "4179783  ID=UTR5:ENST00000460621.6;Parent=ENST000004606...      chrX   \n",
       "4179784  ID=UTR5:ENST00000460621.6;Parent=ENST000004606...      chrX   \n",
       "\n",
       "           start_2      end_2  \n",
       "0           817073     817573  \n",
       "1           822183     822683  \n",
       "2           819940     820440  \n",
       "3           821647     822147  \n",
       "4           825291     825791  \n",
       "...            ...        ...  \n",
       "4179780  155891359  155891859  \n",
       "4179781  155891359  155891859  \n",
       "4179782  155891359  155891859  \n",
       "4179783  155881049  155881549  \n",
       "4179784  155891359  155891859  \n",
       "\n",
       "[4179785 rows x 12 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode = pd.read_table(\"./atac_peaks_filtered.bed\", comment=\"#\",\n",
    "                        sep = \"\\t\", names = ['seqname','source','feature','start','end','NA1','NA2','NA3','attribute','seqname_2', 'start_2' , 'end_2'])\n",
    "gencode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 219399 entries, 0 to 219398\n",
      "Data columns (total 4 columns):\n",
      " #   Column     Non-Null Count   Dtype \n",
      "---  ------     --------------   ----- \n",
      " 0   seqname    219399 non-null  object\n",
      " 1   start_2    219399 non-null  int64 \n",
      " 2   end_2      219399 non-null  int64 \n",
      " 3   attribute  219399 non-null  object\n",
      "dtypes: int64(2), object(2)\n",
      "memory usage: 6.7+ MB\n"
     ]
    }
   ],
   "source": [
    "gencode_genes = gencode[(gencode.feature == \"gene\")][['seqname', 'start_2', 'end_2', 'attribute']].copy().reset_index().drop('index', axis=1) # Extract genes\n",
    "gencode_genes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_info(x):\n",
    "    # Extract gene names\n",
    "#     print(x)\n",
    "    g_name = list(filter(lambda x: 'gene_name' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_type = list(filter(lambda x: 'gene_type' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "#     g_status = list(filter(lambda x: 'gene_status' in x,  x.split(\";\")))[0].split(\"=\")[1]\n",
    "    g_leve = int(list(filter(lambda x: 'level' in x,  x.split(\";\")))[0].split(\"=\")[1])\n",
    "    return (g_name, g_type, g_leve)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start_2</th>\n",
       "      <th>end_2</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>817073</td>\n",
       "      <td>817573</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>RP11-206L10.17</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>822183</td>\n",
       "      <td>822683</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>RP11-206L10.17</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>819940</td>\n",
       "      <td>820440</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>RP11-206L10.17</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>821647</td>\n",
       "      <td>822147</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>RP11-206L10.17</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>825291</td>\n",
       "      <td>825791</td>\n",
       "      <td>ID=ENSG00000230021.10;gene_id=ENSG00000230021....</td>\n",
       "      <td>RP11-206L10.17</td>\n",
       "      <td>transcribed_processed_pseudogene</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219394</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155611127</td>\n",
       "      <td>155611627</td>\n",
       "      <td>ID=ENSG00000168939.12;gene_id=ENSG00000168939....</td>\n",
       "      <td>SPRY3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219395</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155612364</td>\n",
       "      <td>155612864</td>\n",
       "      <td>ID=ENSG00000168939.12;gene_id=ENSG00000168939....</td>\n",
       "      <td>SPRY3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219396</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155632064</td>\n",
       "      <td>155632564</td>\n",
       "      <td>ID=ENSG00000168939.12;gene_id=ENSG00000168939....</td>\n",
       "      <td>SPRY3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219397</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155891359</td>\n",
       "      <td>155891859</td>\n",
       "      <td>ID=ENSG00000124333.16;gene_id=ENSG00000124333....</td>\n",
       "      <td>VAMP7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>219398</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155881049</td>\n",
       "      <td>155881549</td>\n",
       "      <td>ID=ENSG00000124333.16;gene_id=ENSG00000124333....</td>\n",
       "      <td>VAMP7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>219399 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname    start_2      end_2  \\\n",
       "0         chr1     817073     817573   \n",
       "1         chr1     822183     822683   \n",
       "2         chr1     819940     820440   \n",
       "3         chr1     821647     822147   \n",
       "4         chr1     825291     825791   \n",
       "...        ...        ...        ...   \n",
       "219394    chrX  155611127  155611627   \n",
       "219395    chrX  155612364  155612864   \n",
       "219396    chrX  155632064  155632564   \n",
       "219397    chrX  155891359  155891859   \n",
       "219398    chrX  155881049  155881549   \n",
       "\n",
       "                                                attribute       gene_name  \\\n",
       "0       ID=ENSG00000230021.10;gene_id=ENSG00000230021....  RP11-206L10.17   \n",
       "1       ID=ENSG00000230021.10;gene_id=ENSG00000230021....  RP11-206L10.17   \n",
       "2       ID=ENSG00000230021.10;gene_id=ENSG00000230021....  RP11-206L10.17   \n",
       "3       ID=ENSG00000230021.10;gene_id=ENSG00000230021....  RP11-206L10.17   \n",
       "4       ID=ENSG00000230021.10;gene_id=ENSG00000230021....  RP11-206L10.17   \n",
       "...                                                   ...             ...   \n",
       "219394  ID=ENSG00000168939.12;gene_id=ENSG00000168939....           SPRY3   \n",
       "219395  ID=ENSG00000168939.12;gene_id=ENSG00000168939....           SPRY3   \n",
       "219396  ID=ENSG00000168939.12;gene_id=ENSG00000168939....           SPRY3   \n",
       "219397  ID=ENSG00000124333.16;gene_id=ENSG00000124333....           VAMP7   \n",
       "219398  ID=ENSG00000124333.16;gene_id=ENSG00000124333....           VAMP7   \n",
       "\n",
       "                               gene_type  gene_level  \n",
       "0       transcribed_processed_pseudogene           2  \n",
       "1       transcribed_processed_pseudogene           2  \n",
       "2       transcribed_processed_pseudogene           2  \n",
       "3       transcribed_processed_pseudogene           2  \n",
       "4       transcribed_processed_pseudogene           2  \n",
       "...                                  ...         ...  \n",
       "219394                    protein_coding           2  \n",
       "219395                    protein_coding           2  \n",
       "219396                    protein_coding           2  \n",
       "219397                    protein_coding           2  \n",
       "219398                    protein_coding           2  \n",
       "\n",
       "[219399 rows x 7 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_genes[\"gene_name\"], gencode_genes[\"gene_type\"], gencode_genes[\"gene_level\"] = zip(*gencode_genes.attribute.apply(lambda x: gene_info(x)))\n",
    "gencode_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = top_10_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_genes = gencode_genes[gencode_genes['gene_name'].isin(gene_list)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start_2</th>\n",
       "      <th>end_2</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14933</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161214979</td>\n",
       "      <td>161215479</td>\n",
       "      <td>ID=ENSG00000158869.11;gene_id=ENSG00000158869....</td>\n",
       "      <td>FCER1G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14934</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161217205</td>\n",
       "      <td>161217705</td>\n",
       "      <td>ID=ENSG00000158869.11;gene_id=ENSG00000158869....</td>\n",
       "      <td>FCER1G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14935</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161214370</td>\n",
       "      <td>161214870</td>\n",
       "      <td>ID=ENSG00000158869.11;gene_id=ENSG00000158869....</td>\n",
       "      <td>FCER1G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14936</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161216484</td>\n",
       "      <td>161216984</td>\n",
       "      <td>ID=ENSG00000158869.11;gene_id=ENSG00000158869....</td>\n",
       "      <td>FCER1G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14937</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161217802</td>\n",
       "      <td>161218302</td>\n",
       "      <td>ID=ENSG00000158869.11;gene_id=ENSG00000158869....</td>\n",
       "      <td>FCER1G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194340</th>\n",
       "      <td>chr19</td>\n",
       "      <td>35902054</td>\n",
       "      <td>35902554</td>\n",
       "      <td>ID=ENSG00000011600.12;gene_id=ENSG00000011600....</td>\n",
       "      <td>TYROBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194341</th>\n",
       "      <td>chr19</td>\n",
       "      <td>35903583</td>\n",
       "      <td>35904083</td>\n",
       "      <td>ID=ENSG00000011600.12;gene_id=ENSG00000011600....</td>\n",
       "      <td>TYROBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194342</th>\n",
       "      <td>chr19</td>\n",
       "      <td>35909116</td>\n",
       "      <td>35909616</td>\n",
       "      <td>ID=ENSG00000011600.12;gene_id=ENSG00000011600....</td>\n",
       "      <td>TYROBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194343</th>\n",
       "      <td>chr19</td>\n",
       "      <td>35908083</td>\n",
       "      <td>35908583</td>\n",
       "      <td>ID=ENSG00000011600.12;gene_id=ENSG00000011600....</td>\n",
       "      <td>TYROBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194344</th>\n",
       "      <td>chr19</td>\n",
       "      <td>35904765</td>\n",
       "      <td>35905265</td>\n",
       "      <td>ID=ENSG00000011600.12;gene_id=ENSG00000011600....</td>\n",
       "      <td>TYROBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>116 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname    start_2      end_2  \\\n",
       "14933     chr1  161214979  161215479   \n",
       "14934     chr1  161217205  161217705   \n",
       "14935     chr1  161214370  161214870   \n",
       "14936     chr1  161216484  161216984   \n",
       "14937     chr1  161217802  161218302   \n",
       "...        ...        ...        ...   \n",
       "194340   chr19   35902054   35902554   \n",
       "194341   chr19   35903583   35904083   \n",
       "194342   chr19   35909116   35909616   \n",
       "194343   chr19   35908083   35908583   \n",
       "194344   chr19   35904765   35905265   \n",
       "\n",
       "                                                attribute gene_name  \\\n",
       "14933   ID=ENSG00000158869.11;gene_id=ENSG00000158869....    FCER1G   \n",
       "14934   ID=ENSG00000158869.11;gene_id=ENSG00000158869....    FCER1G   \n",
       "14935   ID=ENSG00000158869.11;gene_id=ENSG00000158869....    FCER1G   \n",
       "14936   ID=ENSG00000158869.11;gene_id=ENSG00000158869....    FCER1G   \n",
       "14937   ID=ENSG00000158869.11;gene_id=ENSG00000158869....    FCER1G   \n",
       "...                                                   ...       ...   \n",
       "194340  ID=ENSG00000011600.12;gene_id=ENSG00000011600....    TYROBP   \n",
       "194341  ID=ENSG00000011600.12;gene_id=ENSG00000011600....    TYROBP   \n",
       "194342  ID=ENSG00000011600.12;gene_id=ENSG00000011600....    TYROBP   \n",
       "194343  ID=ENSG00000011600.12;gene_id=ENSG00000011600....    TYROBP   \n",
       "194344  ID=ENSG00000011600.12;gene_id=ENSG00000011600....    TYROBP   \n",
       "\n",
       "             gene_type  gene_level  \n",
       "14933   protein_coding           2  \n",
       "14934   protein_coding           2  \n",
       "14935   protein_coding           2  \n",
       "14936   protein_coding           2  \n",
       "14937   protein_coding           2  \n",
       "...                ...         ...  \n",
       "194340  protein_coding           2  \n",
       "194341  protein_coding           2  \n",
       "194342  protein_coding           2  \n",
       "194343  protein_coding           2  \n",
       "194344  protein_coding           2  \n",
       "\n",
       "[116 rows x 7 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_genes.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['FCER1G', 'HLA-DQA1', 'CCL5', 'HLA-DRB1', 'GZMB', 'ARHGAP24', 'GZMH',\n",
       "       'TYROBP', 'HLA-DPB1', 'GZMA'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_10_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "116"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions = gencode_genes['region'].unique()\n",
    "len(regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_genes['region'] = gencode_genes.apply(lambda row: ':'.join([str(row['seqname']), str(row['start_2'])]) + '-' + str(row['end_2']), axis=1)\n",
    "# (gencode_genes['seqname']+\":\"+str(gencode_genes['start_2'])+\"-\"+str(gencode_genes['end_2']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_genes.to_csv(\"ground_truth_10genes_colin.tsv\",sep =\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start_2</th>\n",
       "      <th>end_2</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_level</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14933</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161214979</td>\n",
       "      <td>161215479</td>\n",
       "      <td>ID=ENSG00000158869.11;gene_id=ENSG00000158869....</td>\n",
       "      <td>FCER1G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:161214979-161215479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14934</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161217205</td>\n",
       "      <td>161217705</td>\n",
       "      <td>ID=ENSG00000158869.11;gene_id=ENSG00000158869....</td>\n",
       "      <td>FCER1G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:161217205-161217705</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14935</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161214370</td>\n",
       "      <td>161214870</td>\n",
       "      <td>ID=ENSG00000158869.11;gene_id=ENSG00000158869....</td>\n",
       "      <td>FCER1G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:161214370-161214870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14936</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161216484</td>\n",
       "      <td>161216984</td>\n",
       "      <td>ID=ENSG00000158869.11;gene_id=ENSG00000158869....</td>\n",
       "      <td>FCER1G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:161216484-161216984</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14937</th>\n",
       "      <td>chr1</td>\n",
       "      <td>161217802</td>\n",
       "      <td>161218302</td>\n",
       "      <td>ID=ENSG00000158869.11;gene_id=ENSG00000158869....</td>\n",
       "      <td>FCER1G</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1:161217802-161218302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194340</th>\n",
       "      <td>chr19</td>\n",
       "      <td>35902054</td>\n",
       "      <td>35902554</td>\n",
       "      <td>ID=ENSG00000011600.12;gene_id=ENSG00000011600....</td>\n",
       "      <td>TYROBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>chr19:35902054-35902554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194341</th>\n",
       "      <td>chr19</td>\n",
       "      <td>35903583</td>\n",
       "      <td>35904083</td>\n",
       "      <td>ID=ENSG00000011600.12;gene_id=ENSG00000011600....</td>\n",
       "      <td>TYROBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>chr19:35903583-35904083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194342</th>\n",
       "      <td>chr19</td>\n",
       "      <td>35909116</td>\n",
       "      <td>35909616</td>\n",
       "      <td>ID=ENSG00000011600.12;gene_id=ENSG00000011600....</td>\n",
       "      <td>TYROBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>chr19:35909116-35909616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194343</th>\n",
       "      <td>chr19</td>\n",
       "      <td>35908083</td>\n",
       "      <td>35908583</td>\n",
       "      <td>ID=ENSG00000011600.12;gene_id=ENSG00000011600....</td>\n",
       "      <td>TYROBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>chr19:35908083-35908583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194344</th>\n",
       "      <td>chr19</td>\n",
       "      <td>35904765</td>\n",
       "      <td>35905265</td>\n",
       "      <td>ID=ENSG00000011600.12;gene_id=ENSG00000011600....</td>\n",
       "      <td>TYROBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2</td>\n",
       "      <td>chr19:35904765-35905265</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>116 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqname    start_2      end_2  \\\n",
       "14933     chr1  161214979  161215479   \n",
       "14934     chr1  161217205  161217705   \n",
       "14935     chr1  161214370  161214870   \n",
       "14936     chr1  161216484  161216984   \n",
       "14937     chr1  161217802  161218302   \n",
       "...        ...        ...        ...   \n",
       "194340   chr19   35902054   35902554   \n",
       "194341   chr19   35903583   35904083   \n",
       "194342   chr19   35909116   35909616   \n",
       "194343   chr19   35908083   35908583   \n",
       "194344   chr19   35904765   35905265   \n",
       "\n",
       "                                                attribute gene_name  \\\n",
       "14933   ID=ENSG00000158869.11;gene_id=ENSG00000158869....    FCER1G   \n",
       "14934   ID=ENSG00000158869.11;gene_id=ENSG00000158869....    FCER1G   \n",
       "14935   ID=ENSG00000158869.11;gene_id=ENSG00000158869....    FCER1G   \n",
       "14936   ID=ENSG00000158869.11;gene_id=ENSG00000158869....    FCER1G   \n",
       "14937   ID=ENSG00000158869.11;gene_id=ENSG00000158869....    FCER1G   \n",
       "...                                                   ...       ...   \n",
       "194340  ID=ENSG00000011600.12;gene_id=ENSG00000011600....    TYROBP   \n",
       "194341  ID=ENSG00000011600.12;gene_id=ENSG00000011600....    TYROBP   \n",
       "194342  ID=ENSG00000011600.12;gene_id=ENSG00000011600....    TYROBP   \n",
       "194343  ID=ENSG00000011600.12;gene_id=ENSG00000011600....    TYROBP   \n",
       "194344  ID=ENSG00000011600.12;gene_id=ENSG00000011600....    TYROBP   \n",
       "\n",
       "             gene_type  gene_level                    region  \n",
       "14933   protein_coding           2  chr1:161214979-161215479  \n",
       "14934   protein_coding           2  chr1:161217205-161217705  \n",
       "14935   protein_coding           2  chr1:161214370-161214870  \n",
       "14936   protein_coding           2  chr1:161216484-161216984  \n",
       "14937   protein_coding           2  chr1:161217802-161218302  \n",
       "...                ...         ...                       ...  \n",
       "194340  protein_coding           2   chr19:35902054-35902554  \n",
       "194341  protein_coding           2   chr19:35903583-35904083  \n",
       "194342  protein_coding           2   chr19:35909116-35909616  \n",
       "194343  protein_coding           2   chr19:35908083-35908583  \n",
       "194344  protein_coding           2   chr19:35904765-35905265  \n",
       "\n",
       "[116 rows x 8 columns]"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_adata = adata[:,top_10_genes]\n",
    "# Assuming you have row and column names associated with your data\n",
    "row_names = subset_adata.obs_names\n",
    "column_names = subset_adata.var_names\n",
    "\n",
    "# Converting the data matrix to a pandas DataFrame with row and column names\n",
    "df_rna = pd.DataFrame(subset_adata.X, index=row_names, columns=column_names)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr1:161214979-161215479',\n",
       " 'chr1:161217205-161217705',\n",
       " 'chr1:161214370-161214870',\n",
       " 'chr1:161216484-161216984',\n",
       " 'chr1:161217802-161218302',\n",
       " 'chr1:161219385-161219885',\n",
       " 'chr1:161220174-161220674',\n",
       " 'chr1:161220772-161221272',\n",
       " 'chr1:161221385-161221885',\n",
       " 'chr1:161222608-161223108',\n",
       " 'chr1:161218412-161218912',\n",
       " 'chr4:85477582-85478082',\n",
       " 'chr4:85482206-85482706',\n",
       " 'chr4:85499502-85500002',\n",
       " 'chr4:85500429-85500929',\n",
       " 'chr4:85498373-85498873',\n",
       " 'chr4:85514135-85514635',\n",
       " 'chr4:85525945-85526445',\n",
       " 'chr4:85531760-85532260',\n",
       " 'chr4:85555137-85555637',\n",
       " 'chr4:85566072-85566572',\n",
       " 'chr4:85603743-85604243',\n",
       " 'chr4:85607565-85608065',\n",
       " 'chr4:85610396-85610896',\n",
       " 'chr4:85657145-85657645',\n",
       " 'chr4:85660519-85661019',\n",
       " 'chr4:85659606-85660106',\n",
       " 'chr4:85669166-85669666',\n",
       " 'chr4:85679255-85679755',\n",
       " 'chr4:85672565-85673065',\n",
       " 'chr4:85681170-85681670',\n",
       " 'chr4:85697826-85698326',\n",
       " 'chr4:85699874-85700374',\n",
       " 'chr4:85701964-85702464',\n",
       " 'chr4:85713020-85713520',\n",
       " 'chr4:85707692-85708192',\n",
       " 'chr4:85717033-85717533',\n",
       " 'chr4:85736496-85736996',\n",
       " 'chr4:85725585-85726085',\n",
       " 'chr4:85733553-85734053',\n",
       " 'chr4:85741260-85741760',\n",
       " 'chr4:85747798-85748298',\n",
       " 'chr4:85743280-85743780',\n",
       " 'chr4:85761766-85762266',\n",
       " 'chr4:85778241-85778741',\n",
       " 'chr4:85796756-85797256',\n",
       " 'chr4:85798344-85798844',\n",
       " 'chr4:85842494-85842994',\n",
       " 'chr4:85863143-85863643',\n",
       " 'chr4:85928892-85929392',\n",
       " 'chr4:85969079-85969579',\n",
       " 'chr4:85491428-85491928',\n",
       " 'chr4:85557213-85557713',\n",
       " 'chr4:85475077-85475577',\n",
       " 'chr4:85671775-85672275',\n",
       " 'chr5:55103930-55104430',\n",
       " 'chr5:55105043-55105543',\n",
       " 'chr5:55102364-55102864',\n",
       " 'chr5:55101430-55101930',\n",
       " 'chr6:32577395-32577895',\n",
       " 'chr6:32578633-32579133',\n",
       " 'chr6:32583456-32583956',\n",
       " 'chr6:32584065-32584565',\n",
       " 'chr6:32580502-32581002',\n",
       " 'chr6:32586296-32586796',\n",
       " 'chr6:32589026-32589526',\n",
       " 'chr6:32589596-32590096',\n",
       " 'chr6:32587701-32588201',\n",
       " 'chr6:32626147-32626647',\n",
       " 'chr6:32627054-32627554',\n",
       " 'chr6:32629187-32629687',\n",
       " 'chr6:32632589-32633089',\n",
       " 'chr6:32627858-32628358',\n",
       " 'chr6:32634210-32634710',\n",
       " 'chr6:32641228-32641728',\n",
       " 'chr6:32639973-32640473',\n",
       " 'chr6:32642545-32643045',\n",
       " 'chr6:32637082-32637582',\n",
       " 'chr6:32638291-32638791',\n",
       " 'chr6:32642009-32642509',\n",
       " 'chr6:32636468-32636968',\n",
       " 'chr6:33076437-33076937',\n",
       " 'chr6:33075595-33076095',\n",
       " 'chr6:33077779-33078279',\n",
       " 'chr6:33084838-33085338',\n",
       " 'chr6:33091119-33091619',\n",
       " 'chr6:33080104-33080604',\n",
       " 'chr6:33081253-33081753',\n",
       " 'chr6:33082641-33083141',\n",
       " 'chr6:33088528-33089028',\n",
       " 'chr6:33080752-33081252',\n",
       " 'chr6:33078803-33079303',\n",
       " 'chr14:24605468-24605968',\n",
       " 'chr14:24607197-24607697',\n",
       " 'chr14:24608233-24608733',\n",
       " 'chr14:24609473-24609973',\n",
       " 'chr14:24629283-24629783',\n",
       " 'chr14:24629919-24630419',\n",
       " 'chr14:24630810-24631310',\n",
       " 'chr14:24632325-24632825',\n",
       " 'chr14:24631747-24632247',\n",
       " 'chr14:24634110-24634610',\n",
       " 'chr17:35870597-35871097',\n",
       " 'chr17:35876371-35876871',\n",
       " 'chr17:35879458-35879958',\n",
       " 'chr17:35871374-35871874',\n",
       " 'chr17:35875497-35875997',\n",
       " 'chr17:35877466-35877966',\n",
       " 'chr17:35880196-35880696',\n",
       " 'chr17:35878216-35878716',\n",
       " 'chr19:35907044-35907544',\n",
       " 'chr19:35902054-35902554',\n",
       " 'chr19:35903583-35904083',\n",
       " 'chr19:35909116-35909616',\n",
       " 'chr19:35908083-35908583',\n",
       " 'chr19:35904765-35905265']"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACACTAATCGTTAGCG-1-10x_pbmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCAATGAAGTAGCGGG-1-10x_pbmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TGAGCTTAGCTTAGCG-1-10x_pbmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GTTTCAGCATGAATAG-1-10x_pbmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACCGAAGCAAAGGCCA-1-10x_pbmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2407</th>\n",
       "      <td>GTTCCCAGTAATCCCT-1-10x_pbmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2408</th>\n",
       "      <td>ATTCACTTCCTAACGG-1-10x_pbmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2409</th>\n",
       "      <td>ATTCGTTTCCAGGAAA-1-10x_pbmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410</th>\n",
       "      <td>AGCCTGGGTAACCAGC-1-10x_pbmc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2411</th>\n",
       "      <td>GTTTAACCATTTGCTC-1-10x_pbmc</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2412 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       orig.ident\n",
       "0     ACACTAATCGTTAGCG-1-10x_pbmc\n",
       "1     GCAATGAAGTAGCGGG-1-10x_pbmc\n",
       "2     TGAGCTTAGCTTAGCG-1-10x_pbmc\n",
       "3     GTTTCAGCATGAATAG-1-10x_pbmc\n",
       "4     ACCGAAGCAAAGGCCA-1-10x_pbmc\n",
       "...                           ...\n",
       "2407  GTTCCCAGTAATCCCT-1-10x_pbmc\n",
       "2408  ATTCACTTCCTAACGG-1-10x_pbmc\n",
       "2409  ATTCGTTTCCAGGAAA-1-10x_pbmc\n",
       "2410  AGCCTGGGTAACCAGC-1-10x_pbmc\n",
       "2411  GTTTAACCATTTGCTC-1-10x_pbmc\n",
       "\n",
       "[2412 rows x 1 columns]"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_adata.var.index = atac_adata.var['features'].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_adata = atac_adata[:,list(regions)]\n",
    "# Assuming you have row and column names associated with your data\n",
    "row_names = subset_adata.obs_names\n",
    "column_names = subset_adata.var_names\n",
    "\n",
    "# Converting the data matrix to a pandas DataFrame with row and column names\n",
    "df_atac = pd.DataFrame(subset_adata.X, index=row_names, columns=column_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_atac.to_csv(\"atac_matrix.csv\")\n",
    "df_rna.to_csv(\"rna_matrix.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "real_scootr",
   "language": "python",
   "name": "real_scootr"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  },
  "vscode": {
   "interpreter": {
    "hash": "eb4bb617a32919d4a1b52f44f72e24e3739c7442a4f58a7bc491ec635ac76433"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
